pkg/verification/verify.go

```diff
@@ -18,17 +18,175 @@ package verification
 import (
 	"bytes"
 	"crypto/x509"
-	"errors"
+	"encoding/asn1"
 	"fmt"
 	"hash"
 	"io"
+	"math/big"
 
 	"github.com/digitorus/pkcs7"
 	"github.com/digitorus/timestamp"
+	"github.com/pkg/errors"
 )
 
+type VerifyOpts struct {
+	Oid            asn1.ObjectIdentifier
+	TsaCertificate *x509.Certificate
+	Intermediates  []*x509.Certificate
+	Roots          []*x509.Certificate
+	Nonce          *big.Int
+	Subject        string
+	HashAlgorithm  hash.Hash
+	HashedMessage  []byte
+}
+
+// Verify the TSR's certificate identifier matches a provided TSA certificate
+func verifyESSCertID(tsaCert *x509.Certificate, opts VerifyOpts) error {
+	if opts.TsaCertificate == nil {
+		return nil
+	}
+
+	errMessage := ""
+
+	if !bytes.Equal(opts.TsaCertificate.RawIssuer, tsaCert.RawIssuer) {
+		errMessage += "TSR cert issuer does not match provided TSA cert issuer"
+	}
+
+	if opts.TsaCertificate.SerialNumber.Cmp(tsaCert.SerialNumber) != 0 {
+		if errMessage != "" {
+			errMessage += ", TSR cert issuer does not match provided TSA cert issuer"
+		} else {
+			errMessage = "TSR cert issuer does not match provided TSA cert issuer"
+		}
+	}
+
+	if errMessage != "" {
+		return errors.New(errMessage)
+	}
+	return nil
+}
+
+// Verify the leaf certificate's subject and/or subject alternative name matches a provided subject
+func verifyLeafCertSubject(cert *x509.Certificate, opts VerifyOpts) error {
```

Comment:
Hm, thinking more about subject verification, I'm not actually sure exactly what OpenSSL is comparing to, but I think we don't need to be as generic.

We can be more precise about what we're verifying. If you call `Subject.String`, you'll get the string representation of a pkix.Name, which will include all parts of a cert subject (common name, organization, state, etc).

I think what's most useful to compare to is the Common Name field (`cert.Subject.CommonName`). This is really only needed in the case where you aren't providing a leaf certificate and using the included cert, just to make sure you've got the expected TSA cert. I'd recommend updating to specify that we're checking the subject's Common Name only, and probably update the VerifyOpts to say CommonName rather than Subject.